guardian/typerighter
# Add prettier to the rule-manager client and to CI

## What does this change?



This adds Prettier to the client app, using the Guardian's [recommended config](https://github.com/guardian/csnx/tree/main/libs/%40guardian/prettier). Files can be formatted with `npm run format:write` and formatting check will run as part of CI. It also runs prettier on client app files.

Instructions have been added to the README on how to configure the Prettier plugin for [IntelliJ](https://www.jetbrains.com/help/idea/prettier.html#ws_prettier_configure) and for [VS Code](https://github.com/prettier/prettier-vscode#format-on-save).

I haven't added a [`.prettierignore`](https://prettier.io/docs/en/ignore.html#ignoring-files-prettierignore) file as by default prettier will ignore files in version control systems directories (".git", ".svn" and ".hg") and `node_modules`.

I initially looked at adding prettier using the pre-commit tool [`git-format-staged`](https://prettier.io/docs/en/precommit.html#option-5-git-format-stagedhttpsgithubcomhallettjgit-format-staged). Because we already have a `pre-commit` hook running the Scala formatter, I tried adding  `git-format-staged` to this existing script (rather than using `husky`). However this proved trickier and more time consuming to set up that expected. So to avoid further potential unforeseen issues with this tool, this PR simply adds prettier to the client and includes instructions on how to configure personal IDEs.

For documentation in case we do choose to add `git-format-staged` in future, this is the command that can be used to run prettier on staged files:
```
./node_modules/.bin/git-format-staged --formatter './node_modules/.bin/prettier --stdin-filepath "{}"' '**/*'
```

Trello card is [here](https://trello.com/c/pKtETRBG/1553-add-prettier-to-typerighter).

## Possible next steps
- add ESLint to CI
- add ESLint with Prettier plugin to a pre-commit hook

## How to test
This can be tested by changing formatting on files (e.g. changing imports do have double instead of single apostrophes or adding extra indentation) and running `npm run format:check` and/or `npm run format:write`.

This can also be tested by following the README instructions to configure your IDE to format on save, making formatting changes and checking that it does indeed format the file.

I tested that `npm run format:check` is running successfully in CI by committing a file with [un-prettified formatting](https://github.com/guardian/typerighter/pull/393/commits/ed9d9a69e950debe830bdf6fe72b1df53db2a402). As expected, this build [failed](https://github.com/guardian/typerighter/actions/runs/5655287874). It then [passed](https://github.com/guardian/typerighter/actions/runs/5656279472) once I'd prettified the file.
